h/api/groups/search.py

```diff
@@ -0,0 +1,44 @@
+def group_filter(request):
+    """Return an Elasticsearch filter for the given user's groups.
+
+    This will filter out annotations that belong to groups that the given user
+    isn't a member of. Annotations belonging to no group and annotations
+    belonging to any group that the user is a member of, will pass through the
+    filter.
+
+    The returned filter is suitable for inserting into an Es query dict.
+    For example:
+
+        query = {
+            "query": {
+                "filtered": {
+                    "filter": group_filter(),
+                    "query": {...}
+                }
+            }
+        }
+
+    :param request: The Pyramid request object
+    :type request: pyramid.request.Request
+
+    """
+    # Annotations that have no 'group' field or a null value for the group
+    # field should pass through the filter.
+    should_clauses = [{'missing': {'field': 'group'}}]
```

Comment:
Annotations that don't belong to any group (or they are in the "Public" group if you like) have no group field, and all users are always allowed to read them (unless they're private of course but that's dealt with separately) so we always want to let them pass.